packages/react-instantsearch-dom-maps/src/Control.js

```diff
@@ -1,65 +1,58 @@
-import React, { Component } from 'react';
+import React from 'react';
 import PropTypes from 'prop-types';
 import { createClassNames, translatable } from 'react-instantsearch-dom';
-import { STATE_CONTEXT } from './Provider';
+import GeoSearchContext from './GeoSearchContext';
 import withGoogleMaps from './withGoogleMaps';
 
 const cx = createClassNames('GeoSearch');
+const ControlPropTypes = {
+  googleMapsInstance: PropTypes.object.isRequired,
+  translate: PropTypes.func.isRequired,
+};
 
-export class Control extends Component {
-  static propTypes = {
-    googleMapsInstance: PropTypes.object.isRequired,
-    translate: PropTypes.func.isRequired,
-  };
+export const Control = ({ googleMapsInstance, translate, context }) => (
+  <div className={cx('control')}>
+    {context.isRefineOnMapMove || !context.hasMapMoveSinceLastRefine ? (
+      <label className={cx('label')}>
+        <input
+          className={cx('input')}
+          type="checkbox"
+          checked={context.isRefineOnMapMove}
+          onChange={context.toggleRefineOnMapMove}
+        />
+        {translate('control')}
+      </label>
+    ) : (
+      <button
+        className={cx('redo')}
+        onClick={() => context.refineWithInstance(googleMapsInstance)}
+      >
+        {translate('redo')}
+      </button>
+    )}
+  </div>
+);
 
-  static contextTypes = {
-    [STATE_CONTEXT]: PropTypes.shape({
-      isRefineOnMapMove: PropTypes.bool.isRequired,
-      toggleRefineOnMapMove: PropTypes.func.isRequired,
-      hasMapMoveSinceLastRefine: PropTypes.bool.isRequired,
-      refineWithInstance: PropTypes.func.isRequired,
-    }).isRequired,
-  };
+Control.propTypes = {
+  ...ControlPropTypes,
+  context: PropTypes.shape({
+    isRefineOnMapMove: PropTypes.bool.isRequired,
+    toggleRefineOnMapMove: PropTypes.func.isRequired,
+    hasMapMoveSinceLastRefine: PropTypes.bool.isRequired,
+    refineWithInstance: PropTypes.func.isRequired,
+  }),
+};
 
-  getStateContext() {
-    return this.context[STATE_CONTEXT];
-  }
+const ControlWrapper = props => (
+  <GeoSearchContext.Consumer>
+    {context => <Control {...props} context={context} />}
```

Comment:
We don't have to provide the prop `context` it's an implementation detail. We can pick and provide which value is really useful for this component.